atom/browser/api/atom_api_window.cc

```diff
@@ -80,6 +100,34 @@ void Window::OnPageTitleUpdated(bool* prevent_default,
   *prevent_default = Emit("page-title-updated", title);
 }
 
+void Window::OnSelectCertificate(
+    net::SSLCertRequestInfo* cert_request_info,
+    scoped_ptr<content::ClientCertificateDelegate> delegate) {
+  v8::Isolate* isolate = v8::Isolate::GetCurrent();
+  v8::Locker locker(isolate);
+  v8::HandleScope handle_scope(isolate);
+  std::vector<mate::Dictionary> cert_info_list;
+  certificate_delegate_.reset(delegate.get());
+
+  for (auto& cert : cert_request_info->client_certs) {
+    std::string encoded_data;
+    if (cert->valid_expiry().is_null() || cert->HasExpired()
+        || !net::X509Certificate::GetPEMEncoded(
+                cert->os_cert_handle(), &encoded_data))
+      continue;
+
+    mate::Dictionary info(isolate, v8::Object::New(isolate));
+    info.Set("data", encoded_data);
+    info.Set("common_name", cert->issuer().common_name);
+    cert_info_list.push_back(info);
+  }
+
+  Emit("select-certificate",
+       cert_request_info->host_and_port.ToString(),
+       cert_info_list,
+       base::Bind(&OnClientCertificateSelected, base::Unretained(this)));
```

Comment:
We should just pass `delegate` to the callback instead of storing it in `BrowserWindow`:

```c++
base::Bind(&OnClientCertificateSelected, delegate);
```

See also:
https://code.google.com/p/chromium/codesearch#chromium/src/base/callback.h&sq=package:chromium&type=cs&l=189